### ISSUE TITLE
issues with Emojis in labels

### ISSUE THREAD

**Issue by DimaVartanian** [TTTAttributedLabel/TTTAttributedLabel#405]
I've tried to distill this down as much as I could and have a pretty basic reproducible case.

I have a completely empty view with this code, using TTTAttributedLabel 1.10.1

```
TTTAttributedLabel *attributedLabel = [[TTTAttributedLabel alloc] initWithFrame:CGRectMake(100, 100, 200, 0)];
attributedLabel.delegate = self;
attributedLabel.font = [UIFont systemFontOfSize:15];
attributedLabel.text = @"Tap here! \U0001F604 \U0001F610";
attributedLabel.linkAttributes = @{NSFontAttributeName : [UIFont boldSystemFontOfSize:15]};

[attributedLabel addLinkToURL:[NSURL URLWithString:@"DOESNOTMATTER"] withRange:[attributedLabel.text rangeOfString:@"Tap here! \U0001F604 \U0001F610"]];;
CGRect frame = attributedLabel.frame;

frame.size.height = [attributedLabel sizeThatFits:attributedLabel.frame.size].height;

// same result using this
// frame.size.height = [TTTAttributedLabel sizeThatFitsAttributedString:attributedLabel.attributedText withConstraints:attributedLabel.frame.size limitedToNumberOfLines:0].height;
attributedLabel.frame = frame;
[self.view addSubview:attributedLabel];
```

The 2 problems I'm seeing are:
1. Top of emojis are cut off at resting state
2. When links are selected, the whole label shifts down a bit (at which point, the bottom of the emojis are cut off)

GIF here:
![tttattributedlabelissue](https://cloud.githubusercontent.com/assets/921505/3275566/0fea6166-f33d-11e3-9eed-7f93a94c5802.gif)

I get the same result using `sizeToFit`, `sizeThatFits:` and the `TTTAttributedLabel` method `sizeThatFitsAttributedString:withConstraints:limitedToNumberOfLines:`

Any ideas?

**Comment by rehmantalha** [comment]
@qdvictory your solution worked for me using system font line height and setting minimum and maximum line height to my label, I was facing almost same issue but a bit difference that when I have multi line emojis set in my label, emojis overlaps and I was using custom font, however I didn't use lineSpacing attribute 
Thanks

**Comment by bnickel** [comment]
With regards to your second issue of text jumping, I just submitted PR #482 to address this.  The problem is that `activeLinkAttributes` was changing the paragraph styling of the link.

**Comment by DimaVartanian** [comment]
I'm seeing another issue now where line spacing in labels with emojis is all messed up. This is what a multiline label with some emojis looks like using system font of size 15 and no other custom attributes (other than a single color) 
![screenshot 2015-04-06 14 41 16](https://cloud.githubusercontent.com/assets/921505/7009833/16a513ee-dc6b-11e4-9746-b215cab37c5c.png)

Anyone have any ideas on how to even out the line spacing? Is there another submitted issue or any workarounds?

**Comment by ytn3rd** [comment]
Had wondered if this was desired effect. This is what messages does in iOS8.2 so it absolutely isn't the desired effect. Tested with both using line breaks and not using line breaks and does the same. Don't currently have an app using latest TTTAttributedLabel so can't play around with it at the moment.
![fullsizerender](https://cloud.githubusercontent.com/assets/904737/7017804/300eaf9a-dd40-11e4-9377-9d3523352e22.jpg)

**Comment by DimaVartanian** [comment]
Yeah this does not happen with the Apple system views (I tried UILabel and UITextView)

**Comment by jhersh** [comment]
I'm having a bit of trouble reproducing these issues. Here's a simple test:

```objc
- (void)testEmojiSizing {
    label.textAlignment = NSTextAlignmentCenter;
    label.numberOfLines = 0;
    label.text = @"Hello world Hello world Hello world Hello world Hello world 👑 \U0001F604 \U0001F610 Hello world 👑 \U0001F604 \U0001F610 Hello world 👑 \U0001F604 \U0001F610 Hello world 👑";
    [label setFrame:CGRectMake(0, 0, 300, 80)];
    FBSnapshotVerifyView(label, nil);
}
```

I don't see any odd line spacing in the result:

![testemojisizing 2x](https://cloud.githubusercontent.com/assets/175139/7946055/631ae25e-0929-11e5-8831-ca027845c3fd.png)

I've also not seen the emoji being cut off when I restrict `numberOfLines` to 1:

![testemojisizing 2x](https://cloud.githubusercontent.com/assets/175139/7946092/921c4a98-0929-11e5-95e2-d55c3175fce7.png)

Does anyone have a more specific reproducible example?

**Comment by DimaVartanian** [comment]
That's the odd part. If you set the text directly as an NSString, it is fine. But if you use an NSAttributedString instead, you can see the problem. Just wrap your string in one, you don't even have to provide any properties

```NSAttributedString *string = [[NSAttributedString alloc] initWithString:@"Hello world Hello world Hello world Hello world Hello world 👑 \U0001F604 \U0001F610 Hello world  Hello world Hello world 👑 \U0001F604 \U0001F610 Hello world 👑 \U0001F604 \U0001F610 Hello world 👑 Hello world Hello world  Hello world Hello world  Hello world Hello world  Hello world Hello world " attributes:@{}];
    attributedLabel.text = string;```

**Comment by jhersh** [comment]
Yup, that'll do it. Thanks @DimaVartanian!

**Comment by HHuiHao** [comment]
[_textLabel setText:attributedText afterInheritingLabelAttributesAndConfiguringWithBlock:^NSMutableAttributedString *(NSMutableAttributedString *mutableAttributedString) {
        return mutableAttributedString;
    }];

**Comment by HHuiHao** [comment]
You use：
1.Calculate the height using the following method
CGSize textLabelSize = [attributedText boundingRectWithSize:CGSizeMake(300, 150) options:NSStringDrawingUsesLineFragmentOrigin context:NULL].size;
2.Do not forget to set the text size
[attributedText addAttribute:NSFontAttributeName value:[UIFont systemFontOfSize:13.0] range:NSMakeRange(0, atMutableAttributedString.length)];

**Comment by yuchensc** [comment]
@DimaVartanian @jhersh  I was able to reproduce the same issue with attributed text. Is there a workaround to get the correct line spacing for attributed string? Thanks a lot!

**Comment by ProjectDent** [comment]
I'm also having this issue, and setting just the text, instead of attributed text, didn't solve the issue. Nor did calling the `setText:afterInheritingLabelAttributesAndConfiguringWithBlock:` method. Right now the issues are multiple:

- Text on lines with emoji are positioned lower down, more line spacing.
- Bottom of emoji are cut off.
- Some text no longer fits.

It doesn't conform to how a normal UILabel appears, and it should. I actually have a UILabel than animates across and is then replaced with the TTTAttributedLabel, but of course there's a slight change in text position now when it becomes the TTTAttributedLabel.

**Comment by mohandev** [comment]
@DimaVartanian 

If you add the following attributes to NSMutableParagraphStyle and set this to attributed string. Then it fixes the linespacing issue when the text contains emojis

   paraStyle.lineHeightMultiple = 1.0;
    paraStyle.minimumLineHeight = font.lineHeight;
    paraStyle.maximumLineHeight = font.lineHeight;

**Comment by muukii** [comment]
This fixed it!

**Comment by bogardon** [comment]
@mohandev thanks for that! FWIT i don't think you need lineHeightMultiple and minimumLineHeight.

**Comment by xilin** [comment]
@mohandev Thanks man, it works!

**Comment by niklassaers** [comment]
Thanks @mohandev :-)

**Comment by LatteCat** [comment]
@mohandev  thanks! ^_^

**Comment by jpsison29** [comment]
Just add textinset

chatTextLabel.textInsets = UIEdgeInsetsMake(1.5, 0, 1.5, 0)

**Comment by at-quangphu** [comment]
@mohandev Thanks a lot!. 😄

**Comment by GuoZhiQiang** [comment]
When i just set `lineHeightMultiple = 1.0;` . The style will be fine. Thanks !
